torch/csrc/jit/passes/quantization.cpp

```diff
@@ -1754,22 +1754,78 @@ std::tuple<at::Tensor, at::Tensor> FoldConvBatchNorm2dHelper::
   return std::make_tuple(new_w, new_b);
 }
 
+bool extractOptionalBNParams(
+    const script::Module& bn,
+    ConvBNParameters& r) {
+  auto graph = bn.get_method("forward").graph();
+  const PatternInfo& pattern_bn = PatternInfo::parse_from_str(R"(
+      graph(%a, %weight, %bias, %running_mean, %running_var,
+          %training, %momentum, %eps, %cudnn_enabled):
+        %bn_out = aten::batch_norm(%a, %weight, %bias, %running_mean,
+            %running_var, %training, %momentum, %eps, %cudnn_enabled)
+        return (%bn_out) )");
+  const Graph& pattern_bn_graph = *pattern_bn.pattern_graph;
+  const auto& bn_vmap = pattern_bn.vmap;
+
+  const auto& matches = findPatternMatches(pattern_bn_graph, *graph);
+  if (matches.size() > 1) {
+    return false;
+  }
+  const auto& match = matches[0];
+  if (bn.hasattr("eps")) {
+    r.bn_eps = bn.attr("eps").toDouble();
+  } else {
+    auto optional_eps = toIValue(match.values_map.at(bn_vmap.at("eps")));
+    if(!optional_eps) {
+      return false;
+    }
+    r.bn_eps = optional_eps.value().toDouble();
+  }
```

Comment:
can you separate the check for attribute and finding constant parameters?

Reply:
I think they are linked. Only if the `attr` does not exist you have Constant there. We can potentially separate but it would just mean getting a constant fails and you will expect some other part of the code to fill that value. I prefer keeping this way. It is more explicit about what is expected in the model, i.e. no attr means there should be a constant else return false.

Reply:
I think this is missing two independent function: populuating parameters for bn and extracting constant values in the pattern in one place, it would be clearer to split these two things. It doesn't influence the logic that you'll only look at the constant values after you  check attributes

Reply:
This function's purpose is to populate weight, bias, eps of ConvBNParams. Populating parameters of BN is dependent on whether the parameters are available as attr or not. If you split out to find whatever values are constant in the pattern then your code will looks something like this:
```
std::vector<c10::optional<IValue>> constants = findBNConstants();
if (bn.hasattr('eps')) {
...
} else {
   optional_esp = constants[eps_index];
}
```
And I dont think this is the right way to do it. It makes it look like as if BN op has some constant values and we are trying to find those. Only when you read the latter part of the code you realize, oh wait, the constants that `findBNConstants` finds can be attributes as well.
This is not clear in my opinion.

Reply:
yeah the purpose for `findBnConstants` is doing one function, that is, to find constant in batchnorm. This doesn't change whether you run it before you populate the parameter for bn or not. writing it in the current way is just mixing the implementation of this function together with populating the parameter of batchnorm

Reply:
cc @ZolotukhinM could you comment on this?

Reply:
I think it's ok to do both things together: in the end we're checking the following:
"Does batchnorm-module have 'eps', 'weight', and 'bias' as either attribute or an inlined to the graph constant?". It is possible to have 'eps' as a constant and 'weight'/'bias' as attributes.

If we split this function into two, then we would need to merge these somehow:
```
valuesFromAttrs = extractOptionalBNParams(...);
// valuesFromAttrs = {/*eps=*/ None, /*weight=*/ W, /*bias=*/ B};

valuesFromGraphConstants = findEPSWeightBiasConstantsOfBN(...);
// valuesFromGraphConstants = {/*eps=*/ 0.1, /*weight=*/ None, /*bias=*/ None};

mergedConstants = merge(valuesFromAttrs, valuesFromGraphConstants);
// mergedConstants = {0.1, W, B};
```

Reply:
Yeah exactly what I felt.